src/Storages/MergeTree/MergeTreeIndexReader.cpp

```diff
@@ -40,41 +41,57 @@ MergeTreeIndexReader::MergeTreeIndexReader(
     MergeTreeData::DataPartPtr part_,
     size_t marks_count_,
     const MarkRanges & all_mark_ranges_,
-    MarkCache * mark_cache,
-    UncompressedCache * uncompressed_cache,
-    MergeTreeReaderSettings settings)
-    : index(index_)
+    MarkCache * mark_cache_,
+    UncompressedCache * uncompressed_cache_,
+    SecondaryIndexCache * secondary_index_cache_,
+    MergeTreeReaderSettings settings_)
+    : index(std::move(index_)), part(std::move(part_)), marks_count(marks_count_)
+    , all_mark_ranges(all_mark_ranges_), mark_cache(mark_cache_)
+    , uncompressed_cache(uncompressed_cache_), secondary_index_cache(secondary_index_cache_)
+    , settings(std::move(settings_))
+{}
+
+void MergeTreeIndexReader::initStreamIfNeeded()
 {
-    auto index_format = index->getDeserializedFormat(part_->getDataPartStorage(), index->getFileName());
+    if (stream)
+        return;
+
+    auto index_format = index->getDeserializedFormat(part->getDataPartStorage(), index->getFileName());
 
     stream = makeIndexReader(
         index_format.extension,
-        index_,
-        part_,
-        marks_count_,
-        all_mark_ranges_,
+        index,
+        part,
+        marks_count,
+        all_mark_ranges,
         mark_cache,
         uncompressed_cache,
         std::move(settings));
     version = index_format.version;
 
-    stream->adjustRightMark(getLastMark(all_mark_ranges_));
+    stream->adjustRightMark(getLastMark(all_mark_ranges));
     stream->seekToStart();
 }
 
 MergeTreeIndexReader::~MergeTreeIndexReader() = default;
 
-void MergeTreeIndexReader::seek(size_t mark)
-{
-    stream->seekToMark(mark);
-}
-
-void MergeTreeIndexReader::read(MergeTreeIndexGranulePtr & granule)
+MergeTreeIndexGranulePtr MergeTreeIndexReader::read(size_t mark)
 {
-    if (granule == nullptr)
-        granule = index->createIndexGranule();
+    auto load_func = [&] {
+        initStreamIfNeeded();
+        if (stream_mark != mark)
```

Comment:
What happens if `mark` is less than `stream_mark`? It wasn't the case before the change, but now the interface of `MergeTreeIndexReader ` allows it.  Should there be an assert?

Reply:
Not sure what you mean, previous interface `MergeTreeIndexReader::seek()` allowed moving backwards too, and it should work fine AFAICT. (Idk whether callers currently do that.)